rfc/rfc-0000-simplify-runnable-use.md

```diff
@@ -0,0 +1,391 @@
+                          storage: 256Mi
+   outputs:
+      url: spec.params[?(@.name=="repository")].value
+      revision: spec.params[?(@.name=="revision")].value
+
+---
+apiVersion: carto.run/v1alpha1
+kind: ClusterSourceTemplate
+metadata:
+   name: source-linter
+spec:
+   template:
+      apiVersion: carto.run/v1alpha1
+      kind: Runnable
+      metadata:
+         name: $(workload.metadata.name)$-linter
+      spec:
+         runTemplateRef:
+            name: md-linting-pipelinerun
+         inputs:
+            repository: $(workload.spec.source.git.url)$
+            revision: $(workload.spec.source.git.revision)$
+         serviceAccountName: pipeline-run-management-sa
+   urlPath: .status.outputs.url
+   revisionPath: .status.outputs.revision
+```
+
+## Proposed example
+Under this proposal, users would only specify one object with a new immutable
+flag.
+
+```yaml
+apiVersion: carto.run/v1alpha1
+kind: ClusterSourceTemplate    # <== Now a source template
+metadata:
+   name: source-linter
+spec:
+   template:
+      apiVersion: tekton.dev/v1beta1
+      kind: PipelineRun
+      metadata:
+         name: $(workload.metadata.name)$-linter      # <== Refers to workload
+      spec:
+         pipelineRef:
+            name: linter-pipeline
+         params:
+            - name: repository
+              value: $(workload.spec.source.git.url)$      # <== Refers to workload
+            - name: revision
+              value: $(workload.spec.source.git.revision)$ # <== Refers to workload
+         workspaces:
+            - name: shared-workspace
+              volumeClaimTemplate:
+                 spec:
+                    accessModes:
+                       - ReadWriteOnce
+                    resources:
+                       requests:
+                          storage: 256Mi
+   urlPath: spec.params[?(@.name=="repository")].value    # <== Refers to path on the immutable object that will be created
+   revisionPath: spec.params[?(@.name=="revision")].value
+   lifecycle: immutable                                         # <== Flag
+```
+
+# How it Works
+[how-it-works]: #how-it-works
+
+When Cartographer processes a template with the immutable flag, rather than
+creating/updating the templated object it will create/update a Runnable and
+ensure that a matching ClusterRunTemplate exists.
+
+What is meant by "ensure that a matching ClusterRunTemplate exists"? For any
+given supply chain template (a cluster scoped object) there will only need
+to be one ClusterRunTemplate. All of the N Runnables created by N workloads
+will reference this one ClusterRunTemplate. Only if the supply chain
+template is updated will this ClusterRunTemplate object need updating.
+
+## Inputs
+The templated object will have two types of fields: hard-coded and templated.
+Hard coded fields can simply be copied directly into the ClusterRunTemplate.
+Templated fields will be translated. Rather than reading from the specified
```

Comment:
In order for this to take place, I imagine we have to limit the scope to jsonpath-based templating as we couldn't do that for `ytt`, is that right?

Reply:
Yes, that's a good call out. I think that is fine, as the goal of this RFC is to provide a convenience method for easy cases, not to replace the direct usage of Runnable and CRT in all cases.